src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs

```diff
@@ -447,14 +447,19 @@ internal TexlBinding ApplyBindingInternal()
                     bool notCoerceToType = false;
                     if (_allowCoerceToType)
                     {
-                        if (this._expectedReturnType != FormulaType.String)
+                        switch (this._expectedReturnType)
                         {
-                            throw new NotImplementedException();
-                        }
-
-                        if (!StringValue.AllowedListConvertToString.Contains(this.ReturnType))
-                        {
-                            notCoerceToType = true;
+                            case StringType:                                
+                                notCoerceToType = !StringValue.AllowedListConvertToString.Contains(this.ReturnType);
+                                break;
+                            case NumberType:
+                                notCoerceToType = !NumberValue.AllowedListConvertToNumber.Contains(this.ReturnType);
+                                break;
+                            case DecimalType:
+                                notCoerceToType = !DecimalValue.AllowedListConvertToDecimal.Contains(this.ReturnType);
+                                break;
+                            default:
+                                throw new NotImplementedException();

```

Comment:
>NotImplementedException [](http://example.com/codeflow?start=42&length=23)

nit: for debuggability, can we update this message to include the type? 
$"Setting ExpectedReturnType to {_expectedReturnType.GetType().FullName} is not implemented";

Reply:
Updated here. thanks